org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/BatchImageBuilder.java

```diff
@@ -348,4 +368,57 @@ protected void storeProblemsFor(SourceFile sourceFile, CategorizedProblem[] prob
 public String toString() {
 	return "batch image builder for:\n\tnew state: " + this.newState; //$NON-NLS-1$
 }
+
+@Override
+public void startBatch() {
+	this.batchMode = true;
+}
+
+@Override
+protected void writeClassFileContents(ClassFile classFile, IFile file, String qualifiedFileName, boolean isTopLevelType,
+		SourceFile compilationUnit) throws CoreException {
+	byte[] content = classFile.getBytes();
+	if (this.batchMode) {
+		if (JavaBuilder.DEBUG) {
+			trace("Batching changed class file " + file.getName());//$NON-NLS-1$
+		}
+		// flush before limit to avoid OOME:
+		if (!this.classContents.isEmpty() && this.classContentsBytesQueued + content.length >= MAX_CLASS_CONTENTS_BYTES_QUEUED) {
+			flushBatch();
+		}
+		this.classContents.put(file, content);
+		this.classContentsBytesQueued += content.length;
+	} else {
+		if (JavaBuilder.DEBUG) {
+			trace("Writing changed class file " + file.getName());//$NON-NLS-1$
+		}
+		file.write(content, true, true, false, null);
+	}
+}
+
+@Override
+public void endBatch() {
+	try {
+		flushBatch();
+	} finally {
+		this.batchMode = false;
+	}
+}
+
+@Override
+public void flushBatch() {
+	try {
+		if (JavaBuilder.DEBUG) {
+			this.classContents.keySet().forEach(file -> trace("Writing changed class file " + file.getName()));//$NON-NLS-1$
+		}
+		ResourcesPlugin.getWorkspace().write(this.classContents, true, true, false, null, WRITER_SERVICE);
+	} catch (CoreException e) {
+		// Already existing class files should not happen:
+		// Duplicate classes get marked earlier with a "The type {} is already defined"
+		Util.log(e, "JavaBuilder handling CoreException"); //$NON-NLS-1$
```

Comment:
`"JavaBuilder handling CoreException"` - better something like `"Failed to write some of the class files: " + this.classContents.keySet()`.

The main question I have here is what happens in the IDE if the files weren't written - the user & JDT will assume that everything is OK after build, however the actual class files would be out-of-date / not existing, causing various unexpected issues. 

Ideally this failure here should be visible to user in some way, just logging is not enough as many users don't have Error log opened at all.

Reply:
I believe `flushBatch()` should declare `throws CoreException` and don't catch it here.

Reply:
i will add a problem marker for the project.

Reply:
I haven't tried, but I'm curious what the builder infrastructure would do if we simply throw exception here?

Reply:
The errorhandling was taken from the existing 
https://github.com/eclipse-jdt/eclipse.jdt.core/blame/ff84b0ea54a020defe593d1e3956701879d0b0f9/org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/builder/AbstractImageBuilder.java#L222

i do not plan to change it.

Reply:
![image](https://github.com/user-attachments/assets/1109c228-9fe6-46ce-b7a9-6fddca30a3a8)